### ISSUE TITLE
Race condition in ProgressBar

### ISSUE THREAD

**Issue by layday** [prompt-toolkit/python-prompt-toolkit#964]
There appears to be a race condition in the `ProgressBar`'s context manager logic where if an exception is raised (or, indeed, if execution finishes) before the `app` has started, `self.app.is_running` will be false, the `app` will not be exited, and the thread will be left to run indefinitely.  You can reproduce this rather trivially (assuming starting a new thread takes as long as it does for me!) with:

```py
from prompt_toolkit.shortcuts import ProgressBar

with ProgressBar():
    pass
```

**Comment by jonathanslenders** [comment]
That looks like a bug. Thank you for reporting!

**Comment by jonathanslenders** [comment]
I can't reproduce it right now, but I understand the issue. I have to think about this.

Note to self:
I'm often getting this exception as well for the above code:
```
Task was destroyed but it is pending!
task: <Task pending coro=<Renderer.wait_for_cpr_responses.<locals>.wait_for_timeout() running at /home/jonathan/git/python-prompt-toolkit/prompt_toolkit/renderer.py:504> wait_for=<Future pending cb=[<TaskWakeupMethWrapper object at 0x7fd14eaf27f8>()]>>
```

**Comment by GaLaKtIkUs** [comment]
> I can't reproduce it right now, but I understand the issue. I have to think about this.
> 
> Note to self:
> I'm often getting this exception as well for the above code:
> 
> ```
> Task was destroyed but it is pending!
> task: <Task pending coro=<Renderer.wait_for_cpr_responses.<locals>.wait_for_timeout() running at /home/jonathan/git/python-prompt-toolkit/prompt_toolkit/renderer.py:504> wait_for=<Future pending cb=[<TaskWakeupMethWrapper object at 0x7fd14eaf27f8>()]>>
> ```

As a workaround I use:
```python
from time import sleep
from prompt_toolkit.shortcuts import ProgressBar

with ProgressBar() as pb:
    for my_iterable in pb(my_iterable):
        do_action(my_iterable)
    sleep(0.1)

**Comment by layday** [comment]
You could delegate closing the app to the daemon thread, which is what I've done [here](https://github.com/layday/instawow/blob/f4a5e85180439918bd9b6566ffc73f3ccd92c814/instawow/utils.py#L176-L208).  I don't know if there's a more CS-y proper-y way to do it but it works.

**Comment by layday** [comment]
I believe this was fixed in 3.0.15, thank you.
